File: Source/Aladdin.CAPI/CAPI/SignData.cs
using System; 
using System.Diagnostics.CodeAnalysis; 

namespace Aladdin.CAPI
{
    ///////////////////////////////////////////////////////////////////////////
	// Выработка подписи данных
    ///////////////////////////////////////////////////////////////////////////
    [SuppressMessage("Microsoft.Design", "CA1063:ImplementIDisposableCorrectly")]
	public abstract class SignData : RefObject, IAlgorithm
	{
        // конструктор
        public SignData() { privateKey = null; } private IPrivateKey privateKey; 

        // деструктор
        protected override void OnDispose()
        {
            // освободить выделенные ресурсы
            RefObject.Release(privateKey); base.OnDispose();
        }
        // используемый личный ключ
        protected IPrivateKey PrivateKey { get { return privateKey; }}

        // алгоритм подписи хэш-значения
        public virtual SignHash SignHashAlgorithm { get { return null; }}

	    // подписать данные
	    public byte[] Sign(IPrivateKey privateKey, IRand rand, byte[] data, int dataOff, int dataLen)
	    {
		    // подписать данные
		    Init(privateKey, rand); Update(data, dataOff, dataLen); return Finish(rand);
	    }
		// инициализировать алгоритм
		public virtual void Init(IPrivateKey privateKey, IRand rand)
        {
            // освободить выделенные ресурсы
            RefObject.Release(this.privateKey); 

            // сохранить переданные параметры
            this.privateKey = RefObject.AddRef(privateKey); 
        }
		// обработать данные
		public abstract void Update(byte[] data, int dataOff, int dataLen);

		// получить подпись данных
        public virtual byte[] Finish(IRand rand)
        {
            // освободить выделенные ресурсы
            RefObject.Release(privateKey); privateKey = null; return null; 
        }
	}
}

File: Source/Aladdin.CAPI.Java/src/aladdin/capi/SignData.java
package aladdin.capi;
import aladdin.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Выработка подписи данных
///////////////////////////////////////////////////////////////////////////
public abstract class SignData extends RefObject implements IAlgorithm
{
    // конструктор
    public SignData() { privateKey = null; } private IPrivateKey privateKey; 
    
    // деструктор
    @Override protected void onClose() throws IOException
    {
        // освободить выделенные ресурсы
        RefObject.release(privateKey); super.onClose();
    }
    // используемый личный ключ
    protected IPrivateKey privateKey() { return privateKey; }
    
    // алгоритм подписи хэш-значения
    public SignHash signHashAlgorithm() { return null; }
    
	// подписать данные
	public final byte[] sign(IPrivateKey privateKey, 
        IRand rand, byte[] data, int dataOff, int dataLen) throws IOException
	{
		// подписать данные
		init(privateKey, rand); update(data, dataOff, dataLen); return finish(rand);
	}
	// инициализировать алгоритм
	public void init(IPrivateKey privateKey, IRand rand) throws IOException
    {
        // освободить выделенные ресурсы
        RefObject.release(this.privateKey); 
        
        // сохранить личный ключ
        this.privateKey = RefObject.addRef(privateKey);
    }
	// обработать данные
	public abstract void update(byte[] data, int dataOff, int dataLen) throws IOException;
	// получить подпись данных
	public byte[] finish(IRand rand) throws IOException
    {
        // освободить выделенные ресурсы
        RefObject.release(this.privateKey); return null; 
    }
}
